Add factoryboy Project and Allocation unit tests

claire-peters opened the pull request:
This PR includes:
- building out of the test_helpers/factory.py module, particularly Project and Allocation-related factories.
- unit test utility functions
- unit tests for Project models and views
- unit tests for Allocation models and views

There are also a couple of minor changes made to the Allocation views:
- remove login_url from AllocationChangeListView so the redirect when not logged in will be to '/user/login?next={page}'
- add success message to AllocationCreateView post request (successful submission of an allocation creation request now includes a corresponding success message)

Looking forward to your comments!

claire-peters commented on coldfront/core/test_helpers/factories.py:
```diff
@@ -30,24 +104,86 @@ class Meta:
+    class Meta:
+        model = ProjectUserStatusChoice
+        django_get_or_create = ('name',)
+    name = 'Active'
```
Here the "name" value is hard-coded instead of taken from a list of possible values because it's the value that's most useful during testing. A different approach was used in ProjectStatusChoiceFactory (line 117), where possible choices were randomly assigned.

claire-peters commented on coldfront/core/test_helpers/factories.py:
```diff
@@ -1,25 +1,99 @@
-from factory import SubFactory
+### Default values and Faker provider setup ###
+
+project_status_choice_names = ['New', 'Active', 'Archived']
```
The hardcoding of the default values here repeats the default values that are listed in project.utils.add_project_status_choices and project.management.commands.add_default_project_choices. It seems like it could be useful to encode default values for model objects in one location in an app-level config.py file or something similar.

claire-peters commented:
Thanks for catching the issues there - the parts for #541 should be in a separate PR, and I'm happy to put off any formatting changes to the user tests module until it can be paired with a more substantive contribution to said module. Just pushed my changes, let me know what you think.

aebruno reviewed (changes requested):
@claire-peters thanks so much for this PR. All looks great. Should all the tests be passing? When I just try running `coldfront test` I'm getting 3 failures:

```
1. FAIL: test_allocationchangelistview_access (coldfront.core.allocation.test_views.AllocationChangeListViewTest)

 AssertionError: '/?next=%2Fallocation%2Fchange-list' != '/user/login?next=%2Fallocation%2Fchange-list'
- /?next=%2Fallocation%2Fchange-list
+ /user/login?next=%2Fallocation%2Fchange-list
?  ++++++++++
 : Response redirected to '/?next=/allocation/change-list', expected '/user/login?next=/allocation/change-list'Expected '/?next=/allocation/change-list' to equal '/user/login?next=/allocation/change-list'.

2. FAIL: test_allocationcreateview_post (coldfront.core.allocation.test_views.AllocationCreateViewTest)
Test POST to the AllocationCreateView
----------------------------------------

AssertionError: False is not true : Couldn't find 'Allocation requested.' in response

3. FAIL: test_allocationcreateview_post_zeroquantity (coldfront.core.allocation.test_views.AllocationCreateViewTest)
Test POST to the AllocationCreateView

AssertionError: False is not true : Couldn't find 'Allocation requested.' in response
```

If these are expected failures I'm happy to merge. If not, perhaps we can get those passing?

claire-peters commented:
Whup, thanks for catching that. Added the changes to make the tests work - specifically, I made it so a success message would appear on the following page after an allocation request has been successfully submitted and I removed the line `login_url = '/'` from AllocationChangeListView. If there are any issues with this, just let me know.

aebruno reviewed (approved):
Looks great. Thanks!
